### DISCUSSION TITLE
What is the correct way of writing the Hamiltonian ?

### DISCUSSION THREAD

**Discussion by DHIMAN002** [QuSpin/QuSpin#511] [Q&A] [answered]
I am trying get the ground state in a particular symmetry sector (Rotational and number conserving) for a two dimensional system. The Hamiltonian is given by,

H=-J &sum;<sub><i,j></sub> [S<sub>i</sub><sup>x</sup> S<sub>j</sub><sup>x</sup> +S<sub>i</sub><sup>y</sup>S<sub>j</sub><sup>y</sup>+S<sub>i</sub><sup>z</sup>S<sub>j</sub><sup>z</sup> ] ......(1)

The Hamiltonian can be re-written as the following,

H=-J &sum;<sub><i,j></sub> [(1/2)S<sub>i</sub><sup>+</sup> S<sub>j</sub><sup>-</sup> +(1/2)S<sub>i</sub><sup>-</sup>S<sub>j</sub><sup>+</sup>+S<sub>i</sub><sup>z</sup>S<sub>j</sub><sup>z</sup> ] .......(2)

I used two of the ways as shown above to create Hamiltonian, but I got different results. Here I have shown the code,

### Creating Hamiltonian using equation-1 and diagonalizing to get ground state:

```
spin_spin_J=[[-J,NN[j][0],NN[j][1]] for j in range(len(NN))]
static=[["xx",spin_spin_J],["yy",spin_spin_J],["zz",spin_spin_J]]
dynamic=[]
#
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H=hamiltonian(static,[],basis=basis,dtype=np.complex64,**no_checks)
E1,V1=H.eigsh(k=1,which="SA",maxiter=1E8)
psi1=V1[:,0]
```
where, NN is a list containing sites on the nearest neighbor bonds.

### Creating Hamiltonian using equation-2 and diagonalizing to get ground state:

```
spin_spin_J=[[-J,NN[j][0],NN[j][1]] for j in range(len(NN))]
spin_spin_J_V2=[[-J/2,NN[j][0],NN[j][1]] for j in range(len(NN))]

static=[["+-",spin_spin_J_V2],["-+",spin_spin_J_V2],["zz",spin_spin_J]]
dynamic=[]
#
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H=hamiltonian(static,[],basis=basis,dtype=np.complex64,**no_checks)
E1,V1=H.eigsh(k=1,which="SA",maxiter=1E8)
psi1=V1[:,0]
```

I checked the eigenvalues given by both the codes and they are different, but if I rewrite the code for eq-2 in the following manner, then the results matches,

### Exception: 

```
spin_spin_J=[[-J,NN[j][0],NN[j][1]] for j in range(len(NN))]
spin_spin_J_V2=[[-J*2,NN[j][0],NN[j][1]] for j in range(len(NN))]

static=[["+-",spin_spin_J_V2],["-+",spin_spin_J_V2],["zz",spin_spin_J]]
dynamic=[]
#
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H=hamiltonian(static,[],basis=basis,dtype=np.complex64,**no_checks)
E1,V1=H.eigsh(k=1,which="SA",maxiter=1E8)
psi1=V1[:,0]
```

**What is the correct way of writing Hamiltonian ?**
All other parts of the code is same with the code in,
https://weinbe58.github.io/QuSpin/examples/user-basis_example0.html#user-basis-example0-label
except one symmetry map, I am using a rotational symmetry and no translational and parity symmetry.

**Comment by weinbe58** [comment]
Nothing here seems incorrect. Can you post the entire code, including how you define the basis?

**Comment by mgbukov** [comment]
@DHIMAN002 did you consider carefully the definition of the "+" and "-" operators in quspin? 

Note that spin bases have the optional argument `pauli` that allows you to use different conventions, see the documentation:

https://weinbe58.github.io/QuSpin/generated/quspin.basis.spin_basis_1d.html#quspin.basis.spin_basis_1d

**Reply by DHIMAN002** [reply]
Hi Marin,
Thanks for the prompt reply.
I used user_basis and the code for the (1) operator definition, (2) operator dictionary and (3) defining the basis is the following:

```
@cfunc(op_sig_32,
	locals=dict(s=int32,n=int32,b=uint32), )
def op(op_struct_ptr,op_str,site_ind,N,args):
	# using struct pointer to pass op_struct_ptr back to C++ see numba Records
	op_struct = carray(op_struct_ptr,1)[0]
	err = 0
	#
	site_ind = N - site_ind - 1 # convention for QuSpin for mapping from bits to sites.
    
    ##DHIMAN: SUMMARY: Here n and s are the same quantity with different conventions.
    ## The n denotes the particle number and s denotes the direction of spin 
    ## at a particular site respectively.
    ## Whereas b is just a mask which is used to perform logical operations on 
    ## the basis state.    
	n = (op_struct.state>>site_ind)&1 # either 0 or 1
	s = (((op_struct.state>>site_ind)&1)<<1)-1 # either -1 or 1
	b = (1<<site_ind)
	#
	if op_str==120: # "x" is integer value 120 = ord("x")
		op_struct.state ^= b

	elif op_str==121: # "y" is integer value 120 = ord("y")
		op_struct.state ^= b
		op_struct.matrix_ele *= 1.0j*s

	elif op_str==43: # "+" is integer value 43 = ord("+")
		if n: op_struct.matrix_ele = 0
		else: op_struct.state ^= b # create spin

	elif op_str==45: # "-" is integer value 45 = ord("-")
		if n: op_struct.state ^= b # destroy spin
		else: op_struct.matrix_ele = 0

	elif op_str==122: # "z" is integer value 120 = ord("z")
		op_struct.matrix_ele *= s

	elif op_str==110: # "n" is integer value 110 = ord("n")
		op_struct.matrix_ele *= n

	elif op_str==73: # "I" is integer value 73 = ord("I")
		pass

	else:
		op_struct.matrix_ele = 0
		err = -1
	#
	return err
op_args=np.array([],dtype=np.uint32)
op_dict = dict(op=op,op_args=op_args)
# create user basis
basis = user_basis(np.uint32,N,op_dict,allowed_ops=set("+-xyznI"),sps=2,pcon_dict=pcon_dict,**maps)
```


For the spin-half case S<sup>+</sup> and S<sup>-</sup> operators produces a factor of 1, that is why I have  op_struct.matrix_ele=1 (or no need to mention as in the code).

**Comment by mgbukov** [comment, marked as answer]
right, so if you look up the usage example:

https://weinbe58.github.io/QuSpin/examples/user-basis_example0.html#user-basis-example0-label

you'll see that we compare to a spin_basis_1d Hamiltonian with `pauli=True`, and you can read off the definition of "+" and "-" from the previous link I sent you. 

The user basis allows you to adjust the prefactors as you find appropriate. 

Did this resolve the issue?

**Reply by DHIMAN002** [reply]
Hi Marin,
Thanks I understood now the issue about the difference between Paulli matrix and Spin-matrix as in the link,
https://weinbe58.github.io/QuSpin/generated/quspin.basis.spin_basis_1d.html#quspin.basis.spin_basis_1d

**But, now I found another issue:**
Somehow Quspin can not assign the matrix elements if I write the Hamiltonian as in equation-2 (in the main question) or in terms of '+-' or '-+'.
To illustrate it, I take the following simple three site Hamiltonian with site-1 and site-2 are connected via XX-Heisenberg interaction, The site-0 is independent. No symmetry is applied ad even no conservation of particles.

### Code in terms of S<sup>x</sup> and S<sup> y</sup>: 
**Code:**
```
print(basis)
static2=[["xx",[[40,1,2]]],["yy",[[40,1,2]]]]
dynamic=[]
#
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H=hamiltonian(static2,[],basis=basis,dtype=np.complex64,**no_checks)
print(H.toarray())
```

**Output:**
```

reference states: 
array index   /   Fock state   /   integer repr. 
     0.         |1 1 1>           7  
     1.         |1 1 0>           6  
     2.         |1 0 1>           5  
     3.         |1 0 0>           4  
     4.         |0 1 1>           3  
     5.         |0 1 0>           2  
     6.         |0 0 1>           1  
     7.         |0 0 0>           0  
[[ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j 20.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j 20.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j 20.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j 20.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j  0.+0.j]]


```
### Code in terms of S<sup>+</sup> and S<sup> -</sup>: 
**Code:**
```
print(basis)
static2=[["+-",[[40,1,2]]],["-+",[[40,1,2]]]]
dynamic=[]
#
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H=hamiltonian(static2,[],basis=basis,dtype=np.complex64,**no_checks)
print(H.toarray())
```

**Output:**
```
reference states: 
array index   /   Fock state   /   integer repr. 
     0.         |1 1 1>           7  
     1.         |1 1 0>           6  
     2.         |1 0 1>           5  
     3.         |1 0 0>           4  
     4.         |0 1 1>           3  
     5.         |0 1 0>           2  
     6.         |0 0 1>           1  
     7.         |0 0 0>           0  
[[0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j]]


```

In the second result there should be a factor of 40 or 20 in the matrix, but the matrix contains only factor one. I guess working with S<sup>x</sup> and S<sup>y</sup> works, but there is a problem with S<sup>+</sup> and S<sup>-</sup>. I have also included the code for the operators here (I also verified the following operators are equivalent to the convention pauli=0 in https://weinbe58.github.io/QuSpin/generated/quspin.basis.spin_basis_1d.html#quspin.basis.spin_basis_1d):

```
@cfunc(op_sig_32,
    locals=dict(s=int32,n=int32,b=uint32), )
def op(op_struct_ptr,op_str,site_ind,N,args):
    # using struct pointer to pass op_struct_ptr back to C++ see numba Records
    op_struct = carray(op_struct_ptr,1)[0]
    err = 0
    #
    site_ind = N - site_ind - 1 # convention for QuSpin for mapping from bits to sites.
    
    ##DHIMAN: SUMMARY: Here n and s are the same quantity with different conventions.
    ## The n denotes the particle number and s denotes the direction of spin 
    ## at a particular site respectively.
    ## Whereas b is just a mask which is used to perform logical operations on 
    ## the basis state.    
    n = (op_struct.state>>site_ind)&1 # either 0 or 1
    s = (((op_struct.state>>site_ind)&1)<<1)-1 # either -1 or 1
    b = (1<<site_ind)
    #
    if op_str==120:#"x" is integer value 120 = ord("x")
        op_struct.state ^= b
        op_struct.matrix_ele *=0.5

    elif op_str==121: # "y" is integer value 120 = ord("y")
        op_struct.state ^= b
        op_struct.matrix_ele *= 0.5*1j*s

    elif op_str==43: # "+" is integer value 43 = ord("+")
        if n:
            op_struct.matrix_ele = 0
        else: 
            op_struct.state ^= b 
            op_struct.matrix_ele = 1# create spin

    elif op_str==45: # "-" is integer value 45 = ord("-")
        if n: 
            op_struct.state ^= b
            op_struct.matrix_ele = 1 # destroy spin
        else: 
            op_struct.matrix_ele = 0

    elif op_str==122: # "z" is integer value 120 = ord("z")
        op_struct.matrix_ele *= 0.5*s

    elif op_str==110: # "n" is integer value 110 = ord("n")
        op_struct.matrix_ele *= n

    elif op_str==73: # "I" is integer value 73 = ord("I")
        pass

    else:
        op_struct.matrix_ele = 0
        err = -1
    #
    return err
op_args=np.array([],dtype=np.uint32)

```


All the other parts of the code is same as in,
https://weinbe58.github.io/QuSpin/examples/user-basis_example0.html#user-basis-example0-label

**Reply by DHIMAN002** [reply]
@mgbukov Thanks Marine, I have resolved the issue.
I made a mistake in the function "op", I defined `op_struct.matrix_ele = 1`,  instead of `op_struct.matrix_ele *= 1`.
That is why, the matrices were wrong.
Thanks a lot for the replies, it helped me a lot to understand.
